app/src/org/commcare/activities/FormEntryActivity.java

```diff
@@ -1112,12 +1116,17 @@ private void handleFormLoadCompletion(AndroidFormController fc) {
 
         formEntryRestoreSession.replaySession(this);
 
+        // jump to form index, no action if null
+        mFormController.returnToStoredIndex();
+
         uiController.refreshView();
         FormNavigationUI.updateNavigationCues(this, mFormController, uiController.questionsView);
         if (isRestartAfterSessionExpiration) {
             Toast.makeText(this,
                     Localization.get("form.entry.restart.after.expiration"), Toast.LENGTH_LONG).show();
         }
+
+        HiddenPreferences.clearInterruptedFormIndex();
```

Comment:
So I'm concerned a bit here about the atomicity and idempotence of how this field is being managed.

It can't be perfect, because of the difference in how the data is stored, but I'm concerned that there are possibilities where this field is set and the other model is removed, and future apps have issues not with the Form Index being null, but with it being misaligned with the form being loaded. I'm also a little concerned that whether the other incomplete form is wiped on recovery could change and because this code isn't close to the recovery/load event it won't be clear that we should change this. 

There are a few ways to potentially improve this, but I think the most straightforward is to stick to the same structure of this PR, but to add one more specific piece, which is adding the form ID (ideally the GUID, but that may not be available easily, so a database ID or some other convenient secondary signal) itself to the item we put into shared preferences, confirming that the form ID matches the recovered form before applying it, and performing this wipe if there's a mismatch to prevent potential problems. 

I'm also not sure if there's an easy or clear way to ensure that the stored FormIndex is valid for the current form, in case something changed in the background, or the form structure is odd somehow (say, there's a section that disappears based on the date), but I'm a little uneasy without the ability to either apply the form index as a wrapped call, or somehow check it before applying the nav.

Reply:
> there are possibilities where this field is set and the other model is removed, and future apps have issues not with the Form Index being null, but with it being misaligned with the form being loaded.
I circled a bit around how much decoupled these should be, envisioning a scenario where they could exist independently. But yeah, there is no point in having a form index that is not directly associated with a session state, so that was me overthinking. One potential candidate ID that could allow some verification during form loading is the SSD record ID.

Reply:
Yeah, SSD record ID is totally fine if that replicates in an identical way before/after, even just the database ID of the incomplete forms row.  I think whatever's convenient and readily available in both the save and load context is great. Just looking for a gut check on the system knowing the FormIndex is actually intended for the recovery context.